File: gateway/src/repositories/target.repository.ts
import { pool } from "../infra/db/pool";

export interface TargetRow {
  id: string;
  user_id: string;
  kind: string;
  model_name: string;
  endpoint_url: string | null;
  encrypted_api_key: string | null;
  load_4_bits: boolean;
  python_target_id: string | null;
  status: string;
  error: string | null;
  created_at: Date;
  updated_at: Date;
}

export interface NewTarget {
  userId: string;
  kind: string;
  modelName: string;
  endpointUrl: string | null;
  encryptedApiKey: string | null;
  load4Bits: boolean;
}

export const targetRepository = {
  async create(t: NewTarget): Promise<TargetRow> {
    const { rows } = await pool.query<TargetRow>(
      `INSERT INTO targets
         (user_id, kind, model_name, endpoint_url, encrypted_api_key, load_4_bits)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [
        t.userId,
        t.kind,
        t.modelName,
        t.endpointUrl,
        t.encryptedApiKey,
        t.load4Bits,
      ],
    );
    return rows[0];
  },

  async listByUser(userId: string): Promise<TargetRow[]> {
    const { rows } = await pool.query<TargetRow>(
      "SELECT * FROM targets WHERE user_id = $1 ORDER BY created_at DESC",
      [userId],
    );
    return rows;
  },

  async findByIdForUser(id: string, userId: string): Promise<TargetRow | null> {
    const { rows } = await pool.query<TargetRow>(
      "SELECT * FROM targets WHERE id = $1 AND user_id = $2",
      [id, userId],
    );
    return rows[0] ?? null;
  },

  async setPythonTargetId(id: string, pythonTargetId: string): Promise<void> {
    await pool.query(
      "UPDATE targets SET python_target_id = $2, updated_at = now() WHERE id = $1",
      [id, pythonTargetId],
    );
  },

  async setStatus(id: string, status: string, error: string | null = null): Promise<void> {
    await pool.query(
      "UPDATE targets SET status = $2, error = $3, updated_at = now() WHERE id = $1",
      [id, status, error],
    );
  },

  async delete(id: string, userId: string): Promise<boolean> {
    const { rowCount } = await pool.query(
      "DELETE FROM targets WHERE id = $1 AND user_id = $2",
      [id, userId],
    );
    return (rowCount ?? 0) > 0;
  },
};

File: apps/service/server/db/target_repository.py
from typing import Any, Dict, List, Optional

from psycopg.rows import dict_row


def _from_row(row: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "target_id": str(row["target_id"]),
        "client_id": row["client_id"],
        "target_key": row["target_key"],
        "config": {
            "kind": row["kind"],
            "model_name": row["model_name"],
            "endpoint_url": row["endpoint_url"],
            "api_key_env": row["api_key_env"],
            "load_4_bits": row["load_4_bits"],
        },
        "created_at": row["created_at"],
    }


class TargetRepository:
    """All target-registration SQL. Bound to one connection."""

    def __init__(self, conn):
        self.conn = conn

    def upsert(self, target_id: str, client_id: str, target_key: str, config: Dict[str, Any]) -> None:
        with self.conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO targets
                    (target_id, client_id, target_key, kind, model_name,
                     endpoint_url, api_key_env, load_4_bits)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (target_id) DO UPDATE SET
                    endpoint_url = EXCLUDED.endpoint_url,
                    api_key_env = EXCLUDED.api_key_env,
                    load_4_bits = EXCLUDED.load_4_bits,
                    updated_at = now()
                """,
                (
                    target_id, client_id, target_key,
                    config.get("kind"), config.get("model_name"),
                    config.get("endpoint_url"), config.get("api_key_env"),
                    bool(config.get("load_4_bits", False)),
                ),
            )

    def get(self, client_id: str, target_id: str) -> Optional[Dict[str, Any]]:
        with self.conn.cursor(row_factory=dict_row) as cur:
            cur.execute(
                "SELECT * FROM targets WHERE target_id = %s AND client_id = %s",
                (target_id, client_id),
            )
            row = cur.fetchone()
            return _from_row(row) if row else None

    def list_for_client(self, client_id: str) -> List[Dict[str, Any]]:
        with self.conn.cursor(row_factory=dict_row) as cur:
            cur.execute(
                "SELECT * FROM targets WHERE client_id = %s ORDER BY created_at",
                (client_id,),
            )
            return [_from_row(row) for row in cur.fetchall()]

    def list_all(self) -> List[Dict[str, Any]]:
        with self.conn.cursor(row_factory=dict_row) as cur:
            cur.execute("SELECT * FROM targets")
            return [_from_row(row) for row in cur.fetchall()]

    def delete(self, client_id: str, target_id: str) -> bool:
        with self.conn.cursor() as cur:
            cur.execute(
                "DELETE FROM targets WHERE target_id = %s AND client_id = %s",
                (target_id, client_id),
            )
            return cur.rowcount > 0
